content/post/2019-05-16-read-the-source.Rmarkdown

```diff
@@ -0,0 +1,73 @@
+
+# How to read the source of a function/package
+
+Sometimes, finding the source of a function might be as easy as writing its name in the console and voilà! you'll get to read the code. Alas, this won't always work (S3 generics, compiled code...). [Jenny Bryan wrote a detailed how-to for each case](https://github.com/jennybc/access-r-source#accessing-r-source), that [Jim Hester automated as an R package, `lookup`](https://github.com/jimhester/lookup#readme), so all you need to do is to learn how to use `lookup`... as well as read its source, of course! :wink:
+
+Let's explore its basic usage.
+
+```r
+library("lookup")
+rhub::check_for_cran
+```
+
+The snippet above will open the source code of the function, check-cran.R. 
+```r
+lookup(body)
+lookup_browse()
+```
+
+This last snippet will first show the body of the `body()` function locally, and then open a browser window pointing at it in the GitHub mirror of R source. 
+
+All in all, `lookup` is a handy package. :ok_hand: How does it work under the hood? We'll let you read its source, but part of the magic is supported by mirrors of R code hosted on GitHub:
+
+* [mirror of R source code](https://github.com/wch/r-source),
+
+* [mirror of CRAN packages](https://github.com/cran), provided by R-hub :sunglasses:,
+
+* [mirror of Bioconductor](https://github.com/Bioconductor/mirror).
+
+# How to search the source
+
+What if you don't know whose source code you'd like to read, i.e. you'd like to see how `sapply()` is used in the wild? That's another case where the code mirrors mentioned previously can help you! All these mirrors are based on GitHub repos, so to use them, you can
+
+* use [GitHub (advanced) search](https://help.github.com/en/articles/about-searching-on-github)
+
+* use [GitHub search via its V3 API](https://developer.github.com/v3/search/) and [sthe `gh` package](https://github.com/r-lib/gh), but beware of rate limits.
+
+* use [GitHub archive](https://www.githubarchive.org/) via [Google BigQuery](https://developers.google.com/bigquery/) e.g. with the [`bigrquery` package](https://github.com/r-dbi/bigrquery).
+
+If you only want to search for code usage on CRAN, you can... use `lookup`!
+
+```r
+lookup::lookup_usage("sapply")
+```
+
+will open https://github.com/search?l=r&q=%22sapply%22+user%3Acran+language%3AR&ref=searchresults&type=Code&utf8=%E2%9C%93, i.e. a GitHub search URL, restricted to R-hub's CRAN source code mirror.
+
+# Conclusion
+
+Reading the R source is useful, and the process of finding it got smoother thanks to the `lookup` package. A piece of the machinery behind it is R-hub's CRAN source code mirror, about which you can find out more in [its source](https://github.com/r-hub/cranatgh/) and in [our docs](https://docs.r-hub.io/#cranatgh). And once you're browsing our docs, stay a while and have a look at our other free services helping R package development!
```

Comment:
no need to link to cranatgh, that's all internal stuff.

Reply:
even in a post about the source? ok :wink: